File: pt6/1262.greatest-sum-divisible-by-three.java
import java.util.Arrays;

class Solution {
    public int maxSumDivThree(int[] nums) {
        int negative = Integer.MIN_VALUE / 4;
        int[] dp = {0, negative, negative};

        for (int num : nums) {
            int[] prev = Arrays.copyOf(dp, 3);
            for (int r = 0; r < 3; r++) {
                int next = (r + num) % 3;
                dp[next] = Math.max(dp[next], prev[r] + num);
            }
        }

        return dp[0];
    }
}

/*
Explanation

dp[r] stores the largest sum seen so far with remainder r modulo 3. For each
number, either skip it or add it to a previous remainder class. Adding num
moves remainder r to (r + num) % 3.

Only three DP states are needed because divisibility by 3 depends only on the
remainder. Copying the previous array prevents using the same number twice in
one iteration.

Edge cases: impossible remainders start very negative; answer can be 0; numbers
already divisible by 3 accumulate in dp[0].

Time complexity: O(n).
Space complexity: O(1).
*/

File: pt6/1262.greatest-sum-divisible-by-three.go
package main

import "math"

func maxSumDivThree(nums []int) int {
	dp := [3]int{0, math.MinInt / 4, math.MinInt / 4}

	for _, num := range nums {
		prev := dp
		for r, total := range prev {
			next := (r + num) % 3
			if total+num > dp[next] {
				dp[next] = total + num
			}
		}
	}

	return dp[0]
}

/*
Explanation

dp[r] stores the largest sum seen so far with remainder r modulo 3. For each
number, either skip it or add it to a previous remainder class. Adding num to a
sum with remainder r moves it to (r+num)%3.

Only three states are needed because divisibility by 3 depends only on the
remainder. Copying dp into prev prevents using the same number more than once
in one iteration.

Edge cases: impossible remainders start as very negative values; answer may be
0; numbers already divisible by 3 accumulate in dp[0].

Time complexity: O(n).
Space complexity: O(1).
*/
